### PR TITLE
feat(minor): accept stdin as an option #4

### PR DESCRIPTION
**Pull request by redstreet**

### CODE REVIEW

File: autobean_format/main.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 from typing import Iterable, Iterator
 from . import formatter, options_lib
 from autobean_refactor import models, parser as parser_lib
+import fileinput
```
Comment (SEIAROTg) [inline autobean_format/main.py:11]: Move forward and have it sorted with other builtin imports?

Comment (SEIAROTg) [inline autobean_format/main.py:11]: Can you join this in other builtin imports in alphabetical order?

File: autobean_format/main.py [hunk 2]
```diff
@@ -42,8 +43,11 @@ def load_files(self, filename: str) -> Iterator[_File]:
             if filename in visited:
                 continue
             visited.add(filename)
-            with open(filename) as f:
-                text = f.read()
+            lines = []
+            with fileinput.input(files=(filename)) as f:
+                for line in f:
```
Comment (SEIAROTg) [inline autobean_format/main.py:48]: Maybe just `text = ''.join(f)`?

File: autobean_format/main.py [hunk 3]
```diff
@@ -42,8 +43,11 @@ def load_files(self, filename: str) -> Iterator[_File]:
             if filename in visited:
                 continue
             visited.add(filename)
-            with open(filename) as f:
-                text = f.read()
+            lines = []
+            with fileinput.input(files=(filename)) as f:
```
Comment (SEIAROTg) [inline autobean_format/main.py:47]: Redundant noop parentheses: `(filename)`

(I guess the intention is singleton tuple but looks like single filename is also acceptable here)

**Comment by SEIAROTg** [comment]
Sorry, please ignore the check failures. Looks like I didn't set them up properly for PR.

**Comment by redstreet** [comment]
All good catches. Fixed.

**Comment by redstreet** [comment]
Also see my commit message about the behavior of output. This is good, expected behavior IMHO, but we can change it if you see an issue.

**Comment by SEIAROTg** [comment]
Thanks! TIL: there is `fileinput` and `nullcontext`.

I'm not sure if the behavior for inplace and recursive is the best option though.

The include support seems to be taking the advantage that `-` is interpreted as a file named `-` (hacky!) and therefore assuming the include paths are relative to cwd. I feel this assumption may not be obvious especially given the content could be passed through some editor, whose cwd may not match the actual file path. Also, considering the use case of reading from stdin, if we are ok with the formatter reading included files by path, why don't we just let it read the main file by path as well?

Regarding inplace, reading from one place (stdin) and writing it to a different place (stdout) doesn't fit the definition of "inplace" in my opinion. If we are also supporting recursive, then reading from multiple files and dumping the results all to stdout feels even less "inplace". Since that behavior can be achieved with `--output=print`, maybe we can just disallow that combination?

If there isn't a decent use case / behavior, maybe we should just let it error out. What do you think?

**Comment by redstreet** [comment]
> The include support seems to be taking the advantage that `-` is interpreted as a file named `-` (hacky!) and therefore assuming the include paths are relative to cwd. I feel this assumption may not be obvious especially given the content could be passed through some editor, whose cwd may not match the actual file path.

Agree, both good points.

BTW, it's not clear if you are aware, but using `-` to mean `stdin` is a standard convention of sorts in the Unix world. `man cat` for example shows:
```
       With no FILE, or when FILE is -, read standard input.
```

> Also, considering the use case of reading from stdin, if we are ok with the formatter reading included files by path, why don't we just let it read the main file by path as well?

Mainly because a typical use case is to select a few lines in an editor, and format it via `autobean-format`. But your earlier points stand.

> Regarding inplace, reading from one place (stdin) and writing it to a different place (stdout) doesn't fit the definition of "inplace" in my opinion. If we are also supporting recursive, then reading from multiple files and dumping the results all to stdout feels even less "inplace". Since that behavior can be achieved with `--output=print`, maybe we can just disallow that combination?

The current PR's behavior with `inplace` is actually this: only lines read from `stdin` are output on `stdout`. Included files (on `stdin`) will still be written to the files themselves. So we neither read from one place and write to another, nor can this be replaced with `--output=print`.

But again, I think this is moot because your original points are well made.

Given that the typical use case for accepting lines via `stdin` is usually in an editor or through another tool how about we simply disable `--recursive` and any recursive behavior when the input is `stdin`?

**Comment by SEIAROTg** [comment]
> BTW, it's not clear if you are aware, but using - to mean stdin is a standard convention of sorts in the Unix world. man cat for example shows:

Yeah, I'm aware of that. I was trying to understand why include works, and turns out it's because all the includes are treated as relative to a main file `-` in path handling.

> The current PR's behavior with inplace is actually this: only lines read from stdin are output on stdout. Included files (on stdin) will still be written to the files themselves. So we neither read from one place and write to another, nor can this be replaced with --output=print.

Oh yes, you are right.

> Given that the typical use case for accepting lines via stdin is usually in an editor or through another tool how about we simply disable --recursive and any recursive behavior when the input is stdin?

👍

**Comment by redstreet** [comment]
Great, fix submitted.

**Comment by redstreet** [comment]
Awesome, thank you! 🙂

**Pull request merged**